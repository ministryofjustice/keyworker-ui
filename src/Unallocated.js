import React, { Component } from 'react';

class Unallocated extends Component {
  constructor (props) {
    super();
    console.log('in constructor ' + props.jwt);
  }

  render () {
    const offenders = this.props.list.map((a) => {
      return (
        <tr key={a.bookingId}>
          <td className="row-gutters"><a href={a.bookingId}>{a.lastName}, {a.firstName}</a></td>
          <td className="row-gutters">{a.offenderNo}</td>
          <td className="row-gutters">{a.internalLocationDesc}</td>
          <td className="row-gutters">release date todo</td>
          <td className="row-gutters">csra todo</td>
        </tr>
      );
    });
    return (
      <div>
        <div className="pure-u-md-7-12">
          <h1 className="heading-large">Prisoners without key workers</h1>
          <p>These prisoners have not yet been allocated a Key worker. Use the button below to automatically allocate
            them to one. You can override the selections on the next page if you wish.</p>
          <table>
            <thead>
              <tr>
                <th>Name</th>
                <th>NOMS ID</th>
                <th>Location</th>
                <th>RD</th>
                <th>CSRA</th>
              </tr>
            </thead>
            <tbody>{offenders}</tbody>
          </table>
        </div>
        <div>
          <button className="button top-gutter" onClick={() => this.props.gotoNext()}>Allocate</button>
        </div>
      </div>
    );
  }
}

export default Unallocated;
