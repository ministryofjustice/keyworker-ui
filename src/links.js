const getOffenderLink = (offenderNo) => {
  return `https://notm-dev.hmpps.dsd.io/offenders/${offenderNo}/personal`;
};

const links = {
  getOffenderLink
};

module.exports = links;
