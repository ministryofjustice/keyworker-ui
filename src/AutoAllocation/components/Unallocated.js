import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ReactRouterPropTypes from 'react-router-prop-types'
import { properCaseName, renderDate } from '../../stringUtils'
import { getOffenderLink } from '../../links'
import { unallocatedListType } from '../../types'

class Unallocated extends Component {
  buildTableForRender() {
    const { unallocatedList } = this.props

    if (!(unallocatedList && unallocatedList.map)) return []

    return unallocatedList.map(a => {
      const formattedName = `${properCaseName(a.lastName)}, ${properCaseName(a.firstName)}`
      return (
        <tr key={a.offenderNo}>
          <td className="row-gutters">
            <a target="_blank" rel="noopener noreferrer" className="link" href={getOffenderLink(a.offenderNo)}>
              {formattedName}
            </a>
          </td>
          <td className="row-gutters">{a.offenderNo}</td>
          <td className="row-gutters">{a.assignedLivingUnitDesc}</td>
          <td className="row-gutters">{renderDate(a.confirmedReleaseDate)}</td>
          <td className="row-gutters">{a.crsaClassification || '--'}</td>
        </tr>
      )
    })
  }

  render() {
    const { history, gotoNext } = this.props
    const offenders = this.buildTableForRender()

    return (
      <div>
        <div className="pure-u-md-7-12 padding-bottom-40">
          <table>
            <thead>
              <tr>
                <th>Prisoner</th>
                <th>Prison no.</th>
                <th>Location</th>
                <th>RD</th>
                <th>CSRA</th>
              </tr>
            </thead>
            <tbody>{offenders}</tbody>
          </table>
          {offenders.length === 0 && (
            <div className="font-small padding-top-large padding-bottom padding-left">No prisoners found</div>
          )}
        </div>
        <div>
          {offenders.length > 0 && (
            <button type="button" className="button" onClick={() => gotoNext(history)}>
              Allocate
            </button>
          )}
        </div>
      </div>
    )
  }
}

Unallocated.propTypes = {
  unallocatedList: unallocatedListType.isRequired,
  gotoNext: PropTypes.func.isRequired,
  history: ReactRouterPropTypes.history.isRequired,
}

export default Unallocated
