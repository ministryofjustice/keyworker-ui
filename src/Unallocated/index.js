import React, { Component } from 'react';
import { properCaseName } from "../stringUtils";
import PropTypes from 'prop-types';
import { getOffenderLink } from "../links";

class Unallocated extends Component {
  buildTableForRender () {
    if (!(this.props.unallocatedList && this.props.unallocatedList.map)) {
      return [];
    }
    return this.props.unallocatedList.map(a => {
      const formattedName = properCaseName(a.lastName) + ', ' + properCaseName(a.firstName);
      return (
        <tr key={a.bookingId}>
          <td className="row-gutters"><a href={getOffenderLink(a.offenderNo)}>{formattedName}</a></td>
          <td className="row-gutters">{a.offenderNo}</td>
          <td className="row-gutters">{a.assignedLivingUnitDesc}</td>
          <td className="row-gutters">{a.confirmedReleaseDate || '--'}</td>
          <td className="row-gutters">{a.crsaClassification || '--'}</td>
        </tr>
      );
    });
  }

  render () {
    const offenders = this.buildTableForRender();
    return (
      <div>
        <div className="pure-u-md-7-12">
          <h1 className="heading-large">Offenders without key workers</h1>
          <p>These offenders have not yet been allocated a Key worker. Use the button below to automatically allocate
            them to one. You can override the selections on the next page if you wish.</p>
          <table>
            <thead>
              <tr>
                <th>Name</th>
                <th>NOMS ID</th>
                <th>Location</th>
                <th>RD</th>
                <th>CSRA</th>
              </tr>
            </thead>
            <tbody>{offenders}</tbody>
          </table>
          {offenders.length === 0 && <div className="font-small padding-top-large padding-bottom padding-left">No prisoners found</div>}
        </div>
        <div>
          {offenders.length > 0 && <button className="button top-gutter margin-bottom" onClick={() => this.props.gotoNext()}>Allocate</button>}
        </div>
      </div>
    );
  }
}

Unallocated.propTypes = {
  unallocatedList: PropTypes.array,
  gotoNext: PropTypes.func.isRequired
};

export default Unallocated;
