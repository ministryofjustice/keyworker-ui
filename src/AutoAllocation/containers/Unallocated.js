import React, { Component } from 'react'
import PropTypes from 'prop-types'
import axios from 'axios'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import Unallocated from '../components/Unallocated'
import { setUnallocatedList, setMessage, setLoaded } from '../../redux/actions/index'
import ErrorComponent from '../../Error/index'
import Spinner from '../../Spinner/index'

import '../../allocation.scss'

class UnallocatedContainer extends Component {
  constructor(props) {
    super()
    this.gotoManualAllocation = this.gotoManualAllocation.bind(this)
    props.unallocatedListDispatch([])
    props.setLoadedDispatch(false)
  }

  async componentWillMount() {
    const { agencyId, user, history, unallocatedListDispatch, handleError, setLoadedDispatch } = this.props

    try {
      if (!user || !user.writeAccess) {
        history.push('/')
        return
      }
      const list = await this.getUnallocated(agencyId)
      unallocatedListDispatch(list)
    } catch (error) {
      handleError(error)
    }
    setLoadedDispatch(true)
  }

  async getUnallocated(agencyId) {
    const response = await axios.get('/api/unallocated', {
      params: {
        agencyId,
      },
    })
    return response.data
  }

  async gotoManualAllocation(history) {
    history.push(`/provisionalAllocation`)
  }

  render() {
    const { loaded } = this.props

    if (loaded) {
      return (
        <div>
          <ErrorComponent {...this.props} />
          {<Unallocated gotoNext={this.gotoManualAllocation} {...this.props} />}
        </div>
      )
    }
    return <Spinner />
  }
}

UnallocatedContainer.propTypes = {
  error: PropTypes.string,
  handleError: PropTypes.func.isRequired,
  unallocatedList: PropTypes.array,
  agencyId: PropTypes.string.isRequired,
  unallocatedListDispatch: PropTypes.func.isRequired,
  setMessageDispatch: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func,
  loaded: PropTypes.bool,
  user: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
}

const mapStateToProps = state => ({
  unallocatedList: state.unallocated.unallocatedList,
  allocatedKeyworkers: state.allocated.allocatedKeyworkers,
  message: state.app.message,
  agencyId: state.app.user.activeCaseLoadId,
  loaded: state.app.loaded,
})

const mapDispatchToProps = dispatch => ({
  unallocatedListDispatch: list => dispatch(setUnallocatedList(list)),
  setMessageDispatch: message => dispatch(setMessage(message)),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
})

export { UnallocatedContainer }
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(UnallocatedContainer))
