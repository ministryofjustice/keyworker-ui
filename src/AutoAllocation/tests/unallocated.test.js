import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Unallocated from '../components/Unallocated';
import links from "../../links";

Enzyme.configure({ adapter: new Adapter() });

const list = [{
  bookingId: 1,
  lastName: "Rendell",
  firstName: "Steve",
  offenderNo: "ZZ124WX",
  assignedLivingUnitDesc: "L-1-1",
  confirmedReleaseDate: "2020-01-02",
  crsaClassification: "High"
},
{
  bookingId: 2,
  lastName: "Rendell2",
  firstName: "Steve2",
  offenderNo: "ZZ125WX",
  assignedLivingUnitDesc: "L-1-2",
  confirmedReleaseDate: null,
  crsaClassification: null
}];

describe('Unallocated component', () => {
  it('should render list correctly', () => {
    links.notmEndpointUrl = "http://my.testUrl/";
    const component = shallow(<Unallocated loaded unallocatedList={list} gotoNext={() => {}} />);

    expect(component).toMatchSnapshot();
  });

  it('should handle click correctly', async () => {
    let callBack = jest.fn();

    const component = shallow(<Unallocated loaded unallocatedList={[list]} gotoNext={callBack}/>);

    component.find('button').simulate('click');
    expect(callBack.mock.calls.length).toEqual(1);
  });

  it('should omit button and show message when no list 1', async () => {
    const component = shallow(<Unallocated loaded gotoNext={jest.fn()}/>);

    expect(component.find('button')).toHaveLength(0);
    expect(component.find('.font-small').debug()).toMatch('No prisoners found');
  });

  it('should omit button and show message when no list 2', async () => {
    const component = shallow(<Unallocated loaded unallocatedList={[]} gotoNext={jest.fn()}/>);

    expect(component.find('button')).toHaveLength(0);
    expect(component.find('.font-small').debug()).toMatch('No prisoners found');
  });
});
