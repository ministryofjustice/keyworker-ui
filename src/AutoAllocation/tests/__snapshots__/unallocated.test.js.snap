// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unallocated component should render list correctly 1`] = `ShallowWrapper {}`;
